library/ssl_msg.c

```diff
@@ -1134,32 +1144,13 @@ static void mbedtls_ssl_cf_memcpy_if_eq( unsigned char *dst,
                                          size_t len,
                                          size_t c1, size_t c2 )
 {
-    /* diff = 0 if c1 == c2, non-zero otherwise */
-    const size_t diff = c1 ^ c2;
-
-    /* MSVC has a warning about unary minus on unsigned integer types,
-     * but this is well-defined and precisely what we want to do here. */
-#if defined(_MSC_VER)
-#pragma warning( push )
-#pragma warning( disable : 4146 )
-#endif
-
-    /* diff_msb's most significant bit is equal to c1 != c2 */
-    const size_t diff_msb = ( diff | -diff );
-
-    /* diff1 = (c1 != c2) in {0, 1} */
-    const size_t diff1 = diff_msb >> ( sizeof( diff_msb ) * 8 - 1 );
-
-    /* mask = c1 != c2 ? 0xff : 0x00 */
-    const unsigned char mask = (unsigned char) -diff1;
-
-#if defined(_MSC_VER)
-#pragma warning( pop )
-#endif
+    /* mask = c1 == c2 ? 0xff : 0x00 */
+    const size_t equal = mbedtls_ssl_cf_bool_eq( c1, c2 );
+    const unsigned char mask = mbedtls_ssl_cf_mask_from_bit( equal );
```

Comment:
Why not `const size_t mask` to avoid any casting ?

Reply:
I'm afraid if we don't need casting here, we'll need it just below in `dst[i] = ( dst[i] & ~mask ) | ( src[i] & mask );` - if `mask` were a `size_t` then the right-hand-side would be a `size_t` as well, and since we're assigning it to an `unsigned char` I suspect MSVC would require and explicit cast as well. I'd rather have the cast here we're doing a simple function call than on a line that already includes a relatively complex expression.

Reply:
Yes, that's right. Let's keep it that way then.